File: Concurrency Programming/1114. Print in Order/1114. Print in Order.java
class Foo {
    private volatile int state = 0;
    private ReentrantLock fooLock = new ReentrantLock();
    private Condition firstDone = fooLock.newCondition();
    private Condition secondDone = fooLock.newCondition();

    public Foo() {}

    public void first(Runnable printFirst) throws InterruptedException {
        fooLock.lock();
        // printFirst.run() outputs "first". Do not change or remove this line.
        printFirst.run();
        state = 1; //mark state as first done.
        firstDone.signalAll();
        fooLock.unlock();
    }

    public void second(Runnable printSecond) throws InterruptedException {
        fooLock.lock();
        while(state != 1)
            firstDone.await();
        // printSecond.run() outputs "second". Do not change or remove this line.
        printSecond.run();
        state = 2;
        secondDone.signalAll();
        fooLock.unlock();
    }

    public void third(Runnable printThird) throws InterruptedException {
        fooLock.lock();
        while(state != 2)
            secondDone.await();
        // printThird.run() outputs "third". Do not change or remove this line.
        printThird.run();
        fooLock.unlock();
    }
}

File: Concurrency Programming/1114. Print in Order/1114. Print in Order.cpp
#include <mutex>
#include <condition_variable>

class Foo {
private:
    std::mutex mtx;
    std::condition_variable firstDone, secondDone;
    int state = 0;
public:
    void first(function<void()> printFirst) {
        std::lock_guard<std::mutex> guard(mtx);
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        state = 1;
        firstDone.notify_all();
    }

    void second(function<void()> printSecond) {
        std::unique_lock<std::mutex> lk(mtx);
        firstDone.wait(lk, [this](){ return state == 1; });
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        state = 2;
        secondDone.notify_all();
    }

    void third(function<void()> printThird) {
        std::unique_lock<std::mutex> lk(mtx);
        secondDone.wait(lk, [this](){ return state == 2; });
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
    }
};
